File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
let traverse = (node, traversal) => {
    if(node === null)
        return
    traversal.push(node.val)
    traverse(node.left, traversal)
    traverse(node.right, traversal)
}

var preorderTraversal = function(root) {
    let preorderTraversal = []
    traverse(root, preorderTraversal)
    
    return preorderTraversal
};

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public List<Integer> traverse(TreeNode node, List<Integer> traversal) {
        if(node != null) {
            traversal.add(node.val);
            traverse(node.left, traversal);
            traverse(node.right, traversal);
        }
        
        return traversal;
    }
    
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> traversal = new ArrayList<>();
        
        return traverse(root, traversal);
    }
}